protocol/custom-gas-token.md

```diff
@@ -0,0 +1,330 @@
+# Custom Gas Token: Design Doc
+
+| Author | *AgusDuha, Joxes, Skeletor* |
+| --- | --- |
+| Created at | *2025-07-24* |
+| Initial Reviewers | *Mark Tyneway, TBD* |
+| Need Approval From | *TBD* |
+| Status | *In Review* |
+
+## Purpose
+
+Enable OP Stack chains to use an asset other than ETH as their native fee currency by introducing a minimal, flexible, and standardizable Custom Gas Token (CGT) architecture. 
+
+## Summary
+
+The proposed Custom Gas Token upgrade lets any OP Stack chain introduce its native asset as the gas currency with almost no core-code intrusion: a single `isCustomGasToken()` flag turns off ETH transfer flows in all bridging methods, while three new pre-deploys, `NativeAssetLiquidity`, a contract with pre-minted assets, `LiquidityController`, an owner-governed mint/burn router, and `L2CGTBridge`, an reserved address that contains a contract with the API calls, hand supply control to chain governors or authorized “minter” contracts that can plug in anything from ERC-20 converters to third-party bridgind models or emission schedules. Wrapped-asset compatibility is preserved, and the entire system launches at genesis by funding the vault and letting the governor release an initial working balance. Overall, the design keeps the OP Stack lean, token-agnostic, and future-proof while unlocking custom economics and flexibility to make the native asset expressive in its manner.
+
+## Problem Statement + Context
+
+The prior CGT design anchored the gas token to an L1 ERC‑20, hard‑coding bridge paths, and cluttering core contracts with token metadata. These restrictions make it hard to handle the following scenarios properly:
+
+- Integrates with a token that doesn’t live on L1.
+- Launch chains whose native asset does not yet exist or have an ERC20 representation somewhere.
```

Comment:
What restrictions are there on the ERC20? Lots of edge cases to consider https://github.com/d-xo/weird-erc20

Reply:
There are no restrictions on the type of ERC20 because the token bridge is not coupled to the design. A bridge can be built that supports any sort of token, even non standard ones. When we build a token bridge, we will likely make it work with tokens that have 18 decimals and are a standard ERC20 token